### PR TITLE
standard file names for JSON results and arrays of results

### PR DESCRIPTION
**Pull request by mgcam**
Depends on https://github.com/wtsi-npg/npg_tracking/pull/549

### CODE REVIEW

File: Changes [hunk 1]
```diff
@@ -1,6 +1,19 @@
 LIST OF CHANGES FOR NPG-QC PACKAGE
 
   - interop parser - extra tests and a fix some result hash keys
```
Comment (dkj) [inline Changes]:
```suggestion
 - interop parser - extra tests and a fix some result hash keys
```

File: Changes [hunk 2]
```diff
@@ -1,6 +1,19 @@
...
+ - provisions for having an array of result objects in the result
+   attribute of the autoqc check object
+ - using moniker role to generate file name root in a standard way
+   from a composition - needed if the result attribute is an array
+   since passing the file name root to the check will not be the right
+   thing to do (multiple file names are needed in this case, so there
+   is no option but delegation it to teh result object itself)
```
Comment (dkj) [inline Changes]:
```suggestion
   is no option but delegation it to the result object itself)
```

File: lib/npg_qc/autoqc/checks/check.pm [hunk 3]
```diff
@@ -42,13 +43,28 @@ npg_qc::autoqc::checks::check
...
 
 =head1 DESCRIPTION
 
 A parent class for autoqc checks. Checks are performed either for a lane or
 for a plex (index, lanelet) or for a composition of the former defined by the
 rpt_list attribute.
 
+The child class is expected to implement a no-argument execute method,
+which performs any necessary computation and sets attributes of either one
+or multiple result objects which should inherit from the npg_qc::autoqc::result::result
```
Comment (dkj) [inline lib/npg_qc/autoqc/checks/check.pm]:
```suggestion
or multiple result objects which should inherit from the
```

File: lib/npg_qc/autoqc/results/result.pm [hunk 4]
```diff
@@ -51,7 +47,10 @@ npg_qc::autoqc::results::result
 
 =head1 DESCRIPTION
 
-A base class to wrap the result of autoqc.
+A class to wrap old-style results of autoqc.
+Uses npg_qc::autoqc::results::base as a parebt class and inherits
```
Comment (dkj) [inline lib/npg_qc/autoqc/results/result.pm]:
```suggestion
Uses npg_qc::autoqc::results::base as a parent class and inherits
```

File: lib/npg_qc/autoqc/results/result.pm [hunk 5]
```diff
@@ -51,7 +47,10 @@ npg_qc::autoqc::results::result
...
+all attributes ans methods of that class. This class adds optional
```
Comment (dkj) [inline lib/npg_qc/autoqc/results/result.pm]:
```suggestion
all attributes and methods of that class. This class adds optional
```

File: t/60-autoqc-checks-check.t [hunk 6]
```diff
@@ -200,14 +200,22 @@ subtest 'input and output directories' => sub {
...
+        'correct listing of qc out directories in the error message';
 
     my $check;
     lives_ok { $check = npg_qc::autoqc::checks::check->new(
                             position => 1,
                             qc_in    => $tdir,
                             id_run   => $idrun)
     } 'object constructor - qc_out is optional';
-    is($check->qc_out, $tdir, 'qc out is set to qc_in');
+    is(scalar @{$check->qc_out}, 1, 'one qc_out directory is asigned');
```
Comment (dkj) [inline t/60-autoqc-checks-check.t]:
```suggestion
    is(scalar @{$check->qc_out}, 1, 'one qc_out directory is assigned');
```

**Comment by mgcam** [comment]
On 13/02/2020 14:48, dkj wrote:

> *@dkj* requested changes on this pull request.
>
All changes done, copyright notice changed using (git log -- 
lib/npg_qc/utils/genotype_calling.pm | grep 'Date:' | cut -d ':' -f 4 | 
cut -d ' ' -f 2 | uniq | sort -g | xargs echo | perl -nle '$_=~s/\ /,/g; 
print "Copyright (C) $_ Genome Research Ltd."' )
>



-- 
 The Wellcome Sanger Institute is operated by Genome Research 
 Limited, a charity registered in England with number 1021457 and a 
 company registered in England with number 2742969, whose registered 
 office is 215 Euston Road, London, NW1 2BE.

**Pull request merged**